### PR TITLE
Remote debugging

### PR DESCRIPTION
**Pull request by romulogoncalves**
The pull request is to provide information how to set Spark for debugging and how to collect info about the Garbage Collector, important when debugging Spark jobs.

Next to it we would provide information on how an user can install extra stuff in the cluster using pip and apt-get. This is handy when we are working on Jupyter notebooks and we have a dependency. The same holds for a Python library that can't be installed using pip or apt-get but you need to have it at the cluster to import it to your notebook.

All the rest are minor things collected over the past 2 months. The several merges you see are because I create the branch (by mistake) from phenology branch instead of the master branch.

### CODE REVIEW

File: roles/spark/templates/spark-defaults.conf.j2 [hunk 1]
```diff
@@ -7,7 +7,8 @@
 # spark.eventLog.dir               hdfs://namenode:8021/directory
 # spark.serializer                 org.apache.spark.serializer.KryoSerializer
 # spark.driver.memory              5g
-# spark.executor.extraJavaOptions  -XX:+PrintGCDetails -Dkey=value -Dnumbers="one two three"
+spark.executor.extraJavaOptions -XX:{{ gc_type }} -XX:+UseCompressedOops {{ gc_debug }} -agentlib:jdwp=transport=dt_socket,server=y,suspend={{ worker_waiting_on_startup }},address={{ worker_debug_port }}
```
Comment (sverhoeven) [inline roles/spark/templates/spark-defaults.conf.j2]: Does this mean that the spark drivers and executors will always run in debug mode? 
If so there could be a performance hit with running in debug mode. 
Maybe comment out the `-agentlib...` and to spark.md#debug mode instruct to uncomment.

File: vars/spark_vars.yml.template [hunk 2]
```diff
@@ -2,9 +2,6 @@
 #Define sbt version
 sbt_version: 0.13.15
 
-#Define Scala release
-scala_release: 2.11
```
Comment (sverhoeven) [inline vars/spark_vars.yml.template]: I still see the var being used in https://github.com/nlesc-sherlock/emma/blob/remote_debugging/roles/spark/tasks/main.yml#L99

So should it be removed?

File: roles/spark/templates/spark-defaults.conf.j2 (hunk 1 above)
Comment (romulogoncalves) [reply roles/spark/templates/spark-defaults.conf.j2]: I have added spark_debug_mode global variable. If you set to true the debug information will be printed and Spark will run in debug mode.

**Comment by romulogoncalves** [comment]
There was an extra commit which is to control the creation of swap space using a global variable. Since I learned how to do it, I decided to do this extra commit.

File: spark.md [hunk 3]
```diff
@@ -31,4 +31,24 @@ SciSpark was installed using the [instructions available at SciSpark GitHub page
 
 A python notebook to do [Unsupervised classification of imagery using scikit-learn](http://nbviewer.jupyter.org/gist/om-henners/c6c8d40389dab75cf535). This example shows how to classify imagery (for example from LANDSAT) using scikit-learn. There are many classification methods available, but for this example they use K-Means as it's simple and fast. It uses [**numpy**](http://www.numpy.org/) and [**rasterio**](https://github.com/mapbox/rasterio).
 
+## Debug mode
+All information to debug set Spark for remote debugging and performance tuning.
 
+### Remote debugging
+To set Spark for remote debugging the user should reconfigure Spark to have only an executor per worker. In **emma/vars/spark_vars/.yml** the user should set *spark_executor_cores* equal to *spark_worker_cores* and *spark_executor_memory* equal to *spark_worker_memory*. Such setting allows us to open a single debugger per executor, i.e., one per node.
```
Comment (sverhoeven) [inline spark.md]: There should be something about `spark_debug_mode` here.  If the mode==False than stuff below does not work. Please describe how to enable Spark debug mode.

Comment (romulogoncalves) [reply spark.md]: I forgot to push that, it is now in.
Good catch

**Pull request merged**